src/gtest/test_unwrap.cpp

```diff
@@ -0,0 +1,350 @@
+        int v = try_unwrap(opt) or_return_unexpected(std::string("err"));
+        return {std::to_string(v)};
+    };
+    EXPECT_EQ(fn(v1), str1);
+    EXPECT_EQ(fn(std::nullopt), err);
+}
+
+struct Custom {
+    static const int ERROR = -1;
+    typedef int value_type;
+    const int val;
+    Custom(int v) : val(v) {}
+    bool has_value() const { return val >= 0; }
+    int value() const { return val; }
+    bool operator==(Custom other) const { return val == other.val; }
+};
+
+//! Test `try_unwrap(...) or_return` for a custom type.
+TEST(Unwrap, Custom) {
+    Custom c1 {1};
+    Custom c2 {2};
+    Custom err {Custom::ERROR};
+
+    auto fn = [&](const Custom& c) -> Custom {
+        int v = try_unwrap(c) or_return;
+        return {v+1};
+    };
+    EXPECT_EQ(fn(c1), c2);
+    EXPECT_EQ(fn(err), err);
+}
+
+struct CustomWrapper {
+    const Custom wrapped;
+    CustomWrapper(const Custom& c) : wrapped(c) {}
+    bool operator==(CustomWrapper other) const { return wrapped == other.wrapped; }
+};
+
+//! Test `try_unwrap(...) or_return` for a custom type with a converting constructor
+//! to a different return type.
+TEST(Unwrap, CustomConvert) {
+    Custom c1 {1};
+    CustomWrapper c2 {Custom(2)};
+    Custom err {Custom::ERROR};
+    CustomWrapper wrapped_err {err};
+
+    auto fn = [&](const Custom& c) -> CustomWrapper {
+        int v = try_unwrap(c) or_return;
+        return {v+1};
+    };
+    EXPECT_EQ(fn(c1), c2);
+    EXPECT_EQ(fn(err), err);
+}
+
+//! Test `try_unwrap(...) or_return_unexpected(...)` for a custom type.
+TEST(Unwrap, CustomReturnUnexpected) {
+    Custom c1 {1};
+    tl::expected<int, std::monostate> ex2 {2};
+    Custom err {Custom::ERROR};
+    tl::expected<int, std::monostate> ex_err {tl::unexpect, std::monostate {}};
+
+    auto fn = [&](const Custom& c) -> tl::expected<int, std::monostate> {
+        int v = try_unwrap(c) or_return_unexpected(std::monostate {});
+        return {v+1};
+    };
+    EXPECT_EQ(fn(c1), ex2);
+    EXPECT_EQ(fn(err), ex_err);
+}
+
+TEST(Unwrap, NonVoidAssertValue) {
+    tl::expected<int, std::string> v1 {1};
+    tl::expected<int, std::string> v2 {2};
+
+    auto fn = [&](const tl::expected<int, std::string>& ex) -> tl::expected<int, std::string> {
+        int v = try_unwrap(ex) or_assert;
+        return {v+1};
+    };
+    EXPECT_EQ(fn(v1), v2);
+}
+
+TEST(Unwrap, VoidAssertValue) {
+    tl::expected<void, std::string> vv {};
+
+    auto fn = [&](const tl::expected<void, std::string>& ex) -> tl::expected<void, std::string> {
+        try_unwrap(ex) or_assert;
+        return {};
+    };
+    EXPECT_TRUE(fn(vv).has_value());
+}
+
+TEST(UnwrapDeathTest, NonVoidAssertError) {
+    tl::expected<int, std::string> err {tl::unexpect, std::string("err")};
+
+    auto fn = [&](const tl::expected<int, std::string>& ex) -> tl::expected<int, std::string> {
+        int v = try_unwrap(ex) or_assert;
+        return {v+1};
+    };
+    ASSERT_DEATH(fn(err), "has_value()");
```

Comment:
I'm not certain these death tests will work on all platforms (in particular, I don't know if Windows will preserve the assertion text that we check for here).